### PR TITLE
 online processing of volumetric data

### PR DESCRIPTION
**Pull request by j-friedrich**
# Description

Main motivation: enable online processing of volumetric data

Side effects: 
- fixed a bunch of previously unnoticed bugs in motion correction, even for 2D data
- added faster OpenCV based gaussian-, maximum- and uniform-filters that support 3D
- added some tests for motion correction and the new filters
- added a small demo for online processing of 3D data 

# Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Changes in documentation or new examples for demos and/or use cases.

# Has your PR been tested?

Yes, 
```python caimanmanager.py test```
and
```python caimanmanager.py demotest```

### CODE REVIEW

File: caiman/base/movies.py [hunk 1]
```diff
@@ -2202,13 +2202,42 @@ def load_iter(file_name, subindices=None, var_name_hdf5: str = 'mov', outtype=np
...
+                                yield vol
+                            t +=1
+                else:
+                    for y in Y:
+                        vol[i] = y.asarray()
+                        i += 1
+                        if i == dims[0]:
+                            i = 0
+                            yield vol
+            elif len(dims)<3 and is3D: # volumetric 3D data w/ 1 volume per file
```
Comment (pgunn) [inline caiman/base/movies.py:2232]: Minor formatting thing: `len(dims)<3` might look better as `len(dims) < 3`

File: caiman/base/movies.py [hunk 2]
```diff
@@ -2202,13 +2202,42 @@ def load_iter(file_name, subindices=None, var_name_hdf5: str = 'mov', outtype=np
...
+                yield load(file_name, subindices=subindices, outtype=outtype, is3D=is3D)
+            else: # 2D data
+                if subindices is not None:
+                    if isinstance(subindices, range):
+                        subindices = slice(subindices.start, subindices.stop, subindices.step)
+                    Y = Y[subindices]
+                for y in Y:
```
Comment (pgunn) [inline caiman/base/movies.py:2239]: It's maybe a bit confusing to use both `y` and `Y` to mean different things

File: caiman/base/rois.py [hunk 3]
```diff
@@ -477,8 +477,8 @@ def register_ROIs(A1,
         if 'csc_matrix' not in str(type(A2)):
             A2 = scipy.sparse.csc_matrix(A2)
 
-        cm_1 = com(A1, dims[0], dims[1])
-        cm_2 = com(A2, dims[0], dims[1])
+        cm_1 = com(A1, *dims)
```
Comment (pgunn) [inline caiman/base/rois.py:480]: nice

File: caiman/source_extraction/cnmf/online_cnmf.py [hunk 4]
```diff
@@ -943,34 +936,31 @@ def initialize_online(self, model_LN=None, T=None):
...
+                        for (x, y) in zip(mc.x_shifts_els, mc.y_shifts_els))
             else:
-                self.estimates.shifts.extend(mc[1])                
+                self.estimates.shifts.extend(mc.shifts_rig)
+            self.min_mov = mc.min_mov
         img_min = Y.min()
 
         if self.params.get('online', 'normalize'):
-            Y -= img_min
+            Y = Y - img_min
```
Comment (pgunn) [inline caiman/source_extraction/cnmf/online_cnmf.py:956]: Both forms are fine; do you prefer this phrasing?

File: caiman/source_extraction/cnmf/online_cnmf.py [hunk 5]
```diff
@@ -1456,9 +1460,9 @@ def bare_initialization(Y, init_batch=1000, k=1, method_init='greedy_roi', gnb=1
         nA = (Ain.power(2).sum(axis=0))
         nr = nA.size
 
-        YA = spdiags(old_div(1., nA), 0, nr, nr) * \
+        YA = spdiags(1./nA, 0, nr, nr) * \
```
Comment (pgunn) [inline caiman/source_extraction/cnmf/online_cnmf.py:1463]: So nice to get rid of old_div finally

File: caiman/source_extraction/cnmf/utilities.py [hunk 6]
```diff
@@ -45,7 +41,163 @@
...
+        if np.isfortran(input):
+            for i in range(input.shape[2]):
+                output[...,i] = cv2.boxFilter(output[...,i], -1, ksize=(size[1],1), borderType=borderType)
+        else:
+            for i in range(input.shape[0]):
+                output[i] = cv2.boxFilter(output[i], -1, ksize=(1,size[1]), borderType=borderType)
+        if mode=='constant' and cval!=0:
+            output+=cval
+        return output
+
```
Comment (pgunn) [inline caiman/source_extraction/cnmf/utilities.py:161]: Should we have an else at the end that throws an exception if ndim is somehow some other value? Feels weird to fall out of the function if the latter elif is false

File: caiman/source_extraction/cnmf/utilities.py [hunk 7]
```diff
@@ -45,7 +41,163 @@
...
+            cv2.MORPH_RECT, size[2::-2]), borderType=borderType)
+        if np.isfortran(input):
+            for i in range(input.shape[2]):
+                output[...,i] = cv2.dilate(output[...,i], cv2.getStructuringElement(
+            cv2.MORPH_RECT, (size[1],1)), borderType=borderType)
+        else:
+            for i in range(input.shape[0]):
+                output[i] = cv2.dilate(output[i], cv2.getStructuringElement(
+            cv2.MORPH_RECT, (1,size[1])), borderType=borderType)
+        return output
```
Comment (pgunn) [inline caiman/source_extraction/cnmf/utilities.py:194]: Same thing with the elif here

File: caiman/source_extraction/cnmf/utilities.py [hunk 8]
```diff
@@ -172,14 +324,14 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
...
+        image_max = ndi.maximum_filter(image, footprint=footprint,
+                                       mode='constant')
+        # image_max = cv2.dilate(image, footprint=footprint, iterations=1)
     else:
         size = 2 * min_distance + 1
-        # image_max = ndi.maximum_filter(image, size=size, mode='constant')
-        image_max = cv2.dilate(image, cv2.getStructuringElement(
-            cv2.MORPH_RECT, (size, size)), iterations=1)
+        image_max = maximum_filter(image, size=size, mode='constant')
+        # image_max = cv2.dilate(image, cv2.getStructuringElement(
```
Comment (pgunn) [inline caiman/source_extraction/cnmf/utilities.py:333]: We may just want to remove the commented-out code rather than update it?

**Comment by pgunn** [comment]
Generally looks like a good diff! Left a few comments

File: caiman/base/movies.py (hunk 1 above)
Comment (j-friedrich) [reply caiman/base/movies.py:2232]: changed it

File: caiman/base/movies.py (hunk 2 above)
Comment (j-friedrich) [reply caiman/base/movies.py:2239]: changed to frame

File: caiman/base/rois.py (hunk 3 above)
Comment (j-friedrich) [reply caiman/base/rois.py:480]: thanks :)

File: caiman/source_extraction/cnmf/utilities.py (hunk 7 above)
Comment (j-friedrich) [reply caiman/source_extraction/cnmf/utilities.py:194]: added an else statement to the end falling back to the scipy.ndimage version

File: caiman/source_extraction/cnmf/utilities.py (hunk 8 above)
Comment (j-friedrich) [reply caiman/source_extraction/cnmf/utilities.py:333]: remove it

**Pull request merged**